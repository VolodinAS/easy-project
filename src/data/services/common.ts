import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { getConfigValue } from '@ijl/cli';
import { Author, CommonResponse, NewProject, NewTask, ProjectId, ProjectType, TaskGetByIdType, TaskId, TaskType } from '../models/common';
import { authToken } from '../../utils/token';

export const apiCommon = createApi({
    reducerPath: 'commonApi',
    baseQuery: fetchBaseQuery({
        baseUrl: getConfigValue('easy-project.api-base-url'),
        prepareHeaders: headers => {
            if (authToken.token) {
                headers.set('Authorization', `Bearer ${authToken.token}`);
            }
            return headers;
        },
    }),
    tagTypes: ['ProjectType', 'TaskType'],
    endpoints: (build) => ({

        getProjects: build.query<CommonResponse<ProjectType[]>, void>({
            query: () => ({
                url: '/projects'
            }),
            providesTags: result => ['ProjectType']
        }),
        getProjectById: build.query<CommonResponse<ProjectType>, { projectId: string }>({
            query: ({projectId}) => ({
                url: `/projects/${projectId}`
            })
        }),
        newProject: build.mutation<CommonResponse<NewProject>, {title: string, code: string, members: object}>({
            query: ({ title, code, members }) => ({
                url: '/projects/new',
                method: 'post',
                body: { title, code, members }
            }),
            invalidatesTags: ['ProjectType']
        }),
        editProject: build.mutation<CommonResponse<NewProject>, {projectId: string, title: string, code: string, members: object}>({
            query: ({ projectId, title, code, members }) => ({
                url: '/projects/edit',
                method: 'post',
                body: { projectId, title, code, members }
            }),
            invalidatesTags: ['ProjectType']
        }),
        deleteProject: build.mutation<CommonResponse<ProjectId>, {projectId: string}>({
            query: (projectId) => ({
                url: '/projects/delete',
                method: 'post',
                body: { projectId }
            }),
            invalidatesTags: ['ProjectType']
        }),



        getTasks: build.query<CommonResponse<TaskType[]>, { projectId: string }>({
            query: ({projectId}) => ({
                url: `/tasks/${projectId}`
            }),
            providesTags: result => ['TaskType']
        }),
        getTaskById: build.query<CommonResponse<TaskType>, { projectId: string, taskId : string}>({
            query: ({ projectId, taskId }) => ({
                url: `/tasks/${projectId}/${taskId}`
            }),
            providesTags: result => ['TaskType']
        }),
        newTask: build.mutation<CommonResponse<NewTask>, { projectId: string, title: string, description: string, type: number, status: number }>({
            query: ({ projectId, title, description, type, status }) => ({
                url: '/tasks/create',
                method: 'post',
                body: { projectId, title, description, type, status }
            }),
            invalidatesTags: ['TaskType']
        }),
        editTask: build.mutation<CommonResponse<NewTask>, { projectId: string, taskId: string, title: string, description: string, type: number, status: number }>({
            query: ({ projectId, taskId, title, description, type, status }) => ({
                url: '/tasks/edit',
                method: 'post',
                body: { projectId, taskId, title, description, type, status }
            }),
            invalidatesTags: ['TaskType']
        }),
        deleteTask: build.mutation<CommonResponse<TaskId>, { projectId: string, taskId: string }>({
            query: ({ projectId, taskId }) => ({
                url: '/tasks/delete',
                method: 'post',
                body: { projectId, taskId }
            }),
            invalidatesTags: ['TaskType']
        }),


        getUsersForSelect: build.query<CommonResponse<Author[]>, void>({
            query: () => ({
                url: '/users/getForSelect',
            }),
        })
    }),
});

export const {
    useGetProjectsQuery,
    useGetProjectByIdQuery,
    useLazyGetProjectByIdQuery,
    useNewProjectMutation,
    useEditProjectMutation,
    useDeleteProjectMutation,
    useGetTasksQuery,
    useGetTaskByIdQuery,
    useLazyGetTaskByIdQuery,

    useNewTaskMutation,
    useEditTaskMutation,
    useDeleteTaskMutation,

    useGetUsersForSelectQuery,
    useLazyGetUsersForSelectQuery,
} = apiCommon;