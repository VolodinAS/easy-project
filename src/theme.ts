const theme = {
    color: {
        typeTask: '#F8BFBF',
        typeFeature: '#B7CFFF',
        typeBug: '#FEFBBD',
        black: '#000000',
        lightPurple: '#CFC7D9',
        purple: '#7A52AA',
        lightGray: '#F5F5F5',
        darkGray: '#9C9C9C',
        red: '#EB0C0C',
        white: '#FFFFFF',
        smallText: '#999999',
        opacityGray: '#ACACAC',
        landingColors: {
            mainGray: '#D5E6EA',
            mainBlue: '#3848A9',
            benefitsItemBackground: '#EDD8BE',
            buttonStartBackground: '#E43958',
            footerBackground: '#343F53',
            headerLogo: '#010101',
            headerBackground: '#F1E7DA'
        },
        dashboardColors: {
            overlayButtonCreate: '#ABE0E7',
            overlayButtonYes: '#B4E7AB',
            overlayButtonNo: '#F9CCBE',
        }
    }
};

export default theme;
