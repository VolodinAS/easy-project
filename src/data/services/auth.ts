import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { getConfigValue } from '@ijl/cli';

import { CommonResponse } from '../models/common';
import { SignInData } from '../models/sing-in';

export const api = createApi({
    reducerPath: 'authApi',
    baseQuery: fetchBaseQuery({
        baseUrl: getConfigValue('easy-project.api-base-url'),
    }),
    tagTypes: [],
    endpoints: (build) => ({
        signIn: build.mutation<CommonResponse<SignInData>, {email: string, password: string}>({
            query: ({ email, password }) => ({
                url: '/sign-in',
                method: 'post',
                body: { email, password }
            })
        }),

        signUp: build.mutation<CommonResponse<unknown>, {email: string, password: string, login: string, name_secondname: string}>({
            query: ({ email, password, login, name_secondname }) => ({
                url: '/registration',
                method: 'post',
                body: { email, password, login, name_secondname }
            })
        })
    }),


});

export const { useSignInMutation, useSignUpMutation } = api;