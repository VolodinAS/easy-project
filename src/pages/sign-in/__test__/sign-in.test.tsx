/* eslint-disable check-file/folder-naming-convention */
import React from 'react';
import { describe, it, expect, beforeAll, jest as jestGlobal, afterAll } from '@jest/globals';
import { fireEvent, render, cleanup, waitFor, getByText, getAllByText } from '@testing-library/react';
import { setupServer } from 'msw/node';
import { rest } from 'msw';
import { ApiProvider } from '@reduxjs/toolkit/dist/query/react';
import { BrowserRouter, useNavigate } from 'react-router-dom';
import { ThemeProvider } from '@emotion/react';

import 'whatwg-fetch';

import theme from '../../../theme';
import { api } from '../../../data/services/auth';
import SignIn from '../index';

import signInSuccess from '../../../../stubs/json/signin/success.json';
import signInFail from '../../../../stubs/json/signin/error.json';

const server = setupServer(
    rest.post('/api/sign-in', (req, res, ctx) => {
        return res(
            ctx.json(signInSuccess)
        );
    })
);

const serverError = setupServer(
    rest.post('/api/sign-in', (req, res, ctx) => {
        return res(
            ctx.status(400),
            ctx.json(signInFail)
        );
    })
);

jestGlobal.mock('react-router-dom', () => {

    const originalModule = jest.requireActual('react-router-dom') as object;

    const navigate = jest.fn();

    return {
        __esModule: true,
        ...originalModule,
        useNavigate: () => navigate
    };
});

describe('Testing SignIn SUCCESS', () => {
    beforeAll( () => {
        jestGlobal.clearAllMocks();
        cleanup();

        server.listen();
    } );
    afterAll( () => {
        server.close();
    } );
    it('check sign-in component render', () => {
        const { container } = render(
            <ThemeProvider theme={theme}>
                <BrowserRouter>
                    <ApiProvider api={api}>
                        <SignIn/>
                    </ApiProvider>
                </BrowserRouter>
            </ThemeProvider>
        );
        expect(container).toMatchSnapshot();
    });

    it('submit empty sign-in form', async () => {
        const errorEmpty = 'Поле не должно быть пустым';
        const { getByTestId, getByLabelText } = render(
            <ThemeProvider theme={theme}>
                <BrowserRouter>
                    <ApiProvider api={api}>
                        <SignIn/>
                    </ApiProvider>
                </BrowserRouter>
            </ThemeProvider>
        );
        const loginForm = getByTestId('signin-form');

        fireEvent.submit(loginForm);

        await waitFor(() => {
            const emptyLabel = getAllByText(loginForm, errorEmpty)[0] as HTMLParagraphElement;
            expect(emptyLabel.textContent).toContain(errorEmpty);
        });
    });

    it('submit sign-in form with only login', async () => {
        const errorEmpty = 'Поле не должно быть пустым';
        const { getByTestId, getByLabelText } = render(
            <ThemeProvider theme={theme}>
                <BrowserRouter>
                    <ApiProvider api={api}>
                        <SignIn/>
                    </ApiProvider>
                </BrowserRouter>
            </ThemeProvider>
        );

        fireEvent.change(getByLabelText(/Почта/i), {
            target: { value: 'alex_vovkin@mail.ru' }
        });

        const loginForm = getByTestId('signin-form');

        fireEvent.submit(loginForm);

        await waitFor(() => {
            const emptyLabel = getAllByText(loginForm, errorEmpty)[0] as HTMLParagraphElement;
            expect(emptyLabel.textContent).toContain(errorEmpty);
        });

        expect(loginForm).toMatchSnapshot();
    });

    it('submit sign-in form with only password', async () => {
        const errorEmpty = 'Поле не должно быть пустым';
        const { getByTestId, getByLabelText } = render(
            <ThemeProvider theme={theme}>
                <BrowserRouter>
                    <ApiProvider api={api}>
                        <SignIn/>
                    </ApiProvider>
                </BrowserRouter>
            </ThemeProvider>
        );

        fireEvent.change(getByLabelText(/Пароль/i), {
            target: { value: '12121992' }
        });

        const loginForm = getByTestId('signin-form');

        fireEvent.submit(loginForm);

        await waitFor(() => {
            const emptyLabel = getAllByText(loginForm, errorEmpty)[0] as HTMLParagraphElement;
            expect(emptyLabel.textContent).toContain(errorEmpty);
        });
        
        expect(loginForm).toMatchSnapshot();
    });

    it('submit sign-in form with incorrect email', async () => {
        const errorInvalid = 'Почта или пароль введены неправильно';
        const { getByTestId, getByLabelText } = render(
            <ThemeProvider theme={theme}>
                <BrowserRouter>
                    <ApiProvider api={api}>
                        <SignIn/>
                    </ApiProvider>
                </BrowserRouter>
            </ThemeProvider>
        );

        fireEvent.change(getByLabelText(/Почта/i), {
            target: { value: 'any email' }
        });

        const loginForm = getByTestId('signin-form');

        fireEvent.submit(loginForm);

        await waitFor(() => {
            const emptyLabel = getAllByText(loginForm, errorInvalid)[0] as HTMLParagraphElement;
            expect(emptyLabel.textContent).toContain(errorInvalid);
        });
        
        expect(loginForm).toMatchSnapshot();
    });

    it('submit sign-in form success', async () => {
        const { getByTestId, getByLabelText } = render(
            <ThemeProvider theme={theme}>
                <BrowserRouter>
                    <ApiProvider api={api}>
                        <SignIn/>
                    </ApiProvider>
                </BrowserRouter>
            </ThemeProvider>
        );

        fireEvent.change(getByLabelText(/Почта/i), {
            target: { value: 'alex_vovkin@mail.ru' }
        });

        fireEvent.change(getByLabelText(/Пароль/i), {
            target: { value: '12121992' }
        });

        const loginForm = getByTestId('signin-form');

        fireEvent.submit(loginForm);

        await waitFor(() => {
            const isSuccess = useNavigate();
            expect(isSuccess).toHaveBeenCalled();
        });
        
        expect(loginForm).toMatchSnapshot();
    });
});


describe('Testing SignIn FAIL', () => {
    beforeAll( () => {
        jestGlobal.clearAllMocks();
        cleanup();

        serverError.listen();
    } );
    afterAll( () => {
        serverError.close();
    } );

    it('submit sign-in form failed', async () => {
        const errorFailed = 'Email или пароль не корректный';
        const { getByTestId, getByLabelText } = render(
            <ThemeProvider theme={theme}>
                <BrowserRouter>
                    <ApiProvider api={api}>
                        <SignIn/>
                    </ApiProvider>
                </BrowserRouter>
            </ThemeProvider>
        );

        fireEvent.change(getByLabelText(/Почта/i), {
            target: { value: 'alex_vovkin@mail.ru' }
        });

        fireEvent.change(getByLabelText(/Пароль/i), {
            target: { value: 'any password' }
        });

        const loginForm = getByTestId('signin-form');

        fireEvent.submit(loginForm);

        await waitFor(() => {
            const errorLabel = getByText(document.body, errorFailed) as HTMLParagraphElement;
            expect(errorLabel.textContent).toContain(errorFailed);
        });
        
        expect(document.body).toMatchSnapshot();
    });
});