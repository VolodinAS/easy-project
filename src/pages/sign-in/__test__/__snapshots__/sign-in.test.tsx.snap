// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SignIn FAIL submit sign-in form failed 1`] = `
<body>
  <div>
    <section
      class="css-zaafzq"
    >
      <section
        class="css-1prn45"
      >
        <button
          class="css-1640pth"
          title="language select"
          value="ru"
        >
          <img
            class="css-zlc08d"
            src="test-file-stub"
          />
        </button>
      </section>
      <form
        class="css-1tcowfc"
        novalidate=""
      >
        <p
          class="css-8vza8i"
        >
          Войти в свой аккаунт
        </p>
        <div
          style="color: red;"
        >
          Email или пароль не корректный
        </div>
        <div
          class="css-xf3brg"
          data-testid="signin-form"
        >
          <div
            class="css-ahns2k"
          >
            <label
              class="css-12e7qyv"
            >
              <span
                class="css-fxi74e"
              >
                Почта
                *
              </span>
              <input
                class="css-d17r1a"
                name="email"
                placeholder="Введите почту..."
                required=""
                type="email"
                value="alex_vovkin@mail.ru"
              />
            </label>
          </div>
          <div
            class="css-ahns2k"
          >
            <label
              class="css-12e7qyv"
            >
              <span
                class="css-fxi74e"
              >
                Пароль
                *
              </span>
              <input
                autocomplete="off"
                class="css-d17r1a"
                name="password"
                placeholder="Введите пароль..."
                required=""
                type="password"
                value="any password"
              />
            </label>
          </div>
        </div>
        <div
          class="css-dpf4m3"
        >
          <p
            class="css-1ewewco"
          >
            Нет аккаунта?
            <a
              class="css-bcl8fr"
              href="/easy-project/registration"
            >
              Зарегистрироваться
            </a>
          </p>
        </div>
        <button
          class="css-1915s5n"
          type="submit"
        >
          Войти
        </button>
      </form>
    </section>
  </div>
</body>
`;

exports[`Testing SignIn SUCCESS check sign-in component render 1`] = `
<div>
  <section
    class="css-zaafzq"
  >
    <section
      class="css-1prn45"
    >
      <button
        class="css-1640pth"
        title="language select"
        value="ru"
      >
        <img
          class="css-zlc08d"
          src="test-file-stub"
        />
      </button>
    </section>
    <form
      class="css-1tcowfc"
      novalidate=""
    >
      <p
        class="css-8vza8i"
      >
        Войти в свой аккаунт
      </p>
      <div
        class="css-xf3brg"
        data-testid="signin-form"
      >
        <div
          class="css-ahns2k"
        >
          <label
            class="css-12e7qyv"
          >
            <span
              class="css-fxi74e"
            >
              Почта
              *
            </span>
            <input
              class="css-d17r1a"
              name="email"
              placeholder="Введите почту..."
              required=""
              type="email"
              value=""
            />
          </label>
        </div>
        <div
          class="css-ahns2k"
        >
          <label
            class="css-12e7qyv"
          >
            <span
              class="css-fxi74e"
            >
              Пароль
              *
            </span>
            <input
              autocomplete="off"
              class="css-d17r1a"
              name="password"
              placeholder="Введите пароль..."
              required=""
              type="password"
              value=""
            />
          </label>
        </div>
      </div>
      <div
        class="css-dpf4m3"
      >
        <p
          class="css-1ewewco"
        >
          Нет аккаунта?
          <a
            class="css-bcl8fr"
            href="/easy-project/registration"
          >
            Зарегистрироваться
          </a>
        </p>
      </div>
      <button
        class="css-1915s5n"
        type="submit"
      >
        Войти
      </button>
    </form>
  </section>
</div>
`;

exports[`Testing SignIn SUCCESS submit sign-in form success 1`] = `
<div
  class="css-xf3brg"
  data-testid="signin-form"
>
  <div
    class="css-ahns2k"
  >
    <label
      class="css-12e7qyv"
    >
      <span
        class="css-fxi74e"
      >
        Почта
        *
      </span>
      <input
        class="css-d17r1a"
        name="email"
        placeholder="Введите почту..."
        required=""
        type="email"
        value="alex_vovkin@mail.ru"
      />
    </label>
  </div>
  <div
    class="css-ahns2k"
  >
    <label
      class="css-12e7qyv"
    >
      <span
        class="css-fxi74e"
      >
        Пароль
        *
      </span>
      <input
        autocomplete="off"
        class="css-d17r1a"
        name="password"
        placeholder="Введите пароль..."
        required=""
        type="password"
        value="12121992"
      />
    </label>
  </div>
</div>
`;

exports[`Testing SignIn SUCCESS submit sign-in form with incorrect email 1`] = `
<div
  class="css-xf3brg"
  data-testid="signin-form"
>
  <div
    class="css-ahns2k"
  >
    <label
      class="css-12e7qyv"
    >
      <span
        class="css-fxi74e"
      >
        Почта
        *
      </span>
      <input
        class="css-d17r1a"
        name="email"
        placeholder="Введите почту..."
        required=""
        type="email"
        value="any email"
      />
    </label>
    <p
      class="css-11un70f"
    >
      Почта или пароль введены неправильно
    </p>
  </div>
  <div
    class="css-ahns2k"
  >
    <label
      class="css-12e7qyv"
    >
      <span
        class="css-fxi74e"
      >
        Пароль
        *
      </span>
      <input
        autocomplete="off"
        class="css-d17r1a"
        name="password"
        placeholder="Введите пароль..."
        required=""
        type="password"
        value=""
      />
    </label>
    <p
      class="css-11un70f"
    >
      Поле не должно быть пустым
    </p>
  </div>
</div>
`;

exports[`Testing SignIn SUCCESS submit sign-in form with only login 1`] = `
<div
  class="css-xf3brg"
  data-testid="signin-form"
>
  <div
    class="css-ahns2k"
  >
    <label
      class="css-12e7qyv"
    >
      <span
        class="css-fxi74e"
      >
        Почта
        *
      </span>
      <input
        class="css-d17r1a"
        name="email"
        placeholder="Введите почту..."
        required=""
        type="email"
        value="alex_vovkin@mail.ru"
      />
    </label>
    
  </div>
  <div
    class="css-ahns2k"
  >
    <label
      class="css-12e7qyv"
    >
      <span
        class="css-fxi74e"
      >
        Пароль
        *
      </span>
      <input
        autocomplete="off"
        class="css-d17r1a"
        name="password"
        placeholder="Введите пароль..."
        required=""
        type="password"
        value=""
      />
    </label>
    <p
      class="css-11un70f"
    >
      Поле не должно быть пустым
    </p>
  </div>
</div>
`;

exports[`Testing SignIn SUCCESS submit sign-in form with only password 1`] = `
<div
  class="css-xf3brg"
  data-testid="signin-form"
>
  <div
    class="css-ahns2k"
  >
    <label
      class="css-12e7qyv"
    >
      <span
        class="css-fxi74e"
      >
        Почта
        *
      </span>
      <input
        class="css-d17r1a"
        name="email"
        placeholder="Введите почту..."
        required=""
        type="email"
        value=""
      />
    </label>
    <p
      class="css-11un70f"
    >
      Поле не должно быть пустым
    </p>
  </div>
  <div
    class="css-ahns2k"
  >
    <label
      class="css-12e7qyv"
    >
      <span
        class="css-fxi74e"
      >
        Пароль
        *
      </span>
      <input
        autocomplete="off"
        class="css-d17r1a"
        name="password"
        placeholder="Введите пароль..."
        required=""
        type="password"
        value="12121992"
      />
    </label>
    
  </div>
</div>
`;
